pkgs/top-level/cuda-packages.nix

```diff
@@ -0,0 +1,142 @@
+    };
+  };
+
+  cudnnVersionForCuda = {
+    "10.0" = "7.4.2";
+    "10.1" = "7.6.3";
+    "10.2" = "7.6.5";
+  }.${cudaVersion} or "8.1.1";
+
+  ### CuTensor
+  
+  buildCuTensorPackage = callPackage ../development/libraries/science/math/cutensor/generic.nix;
+
+  cuTensorVersions = {
+    "1.2.2.5" = {
+      sha256 = "1dl9bd71frhac9cb8lvnh71zfsnqxbxbfhndvva2zf6nh0my4klm";
+    };
+    "1.3.1.3" = {
+      sha256 = "sha256-mNlVnabB2IC3HnYY0mb06RLqQzDxN9ePGVeBy3hkBC8=";
+    };
+  };
+
+  ### Cuda Toolkit Redist
+
+  # Manifest files for redist cudatoolkit
+  cudaToolkitRedistManifests = {
+    "11.4" = ../development/compilers/cudatoolkit/redist/manifests/redistrib_11.4.4.json;
+    "11.5" = ../development/compilers/cudatoolkit/redist/manifests/redistrib_11.5.2.json;
+    "11.6" = ../development/compilers/cudatoolkit/redist/manifests/redistrib_11.6.2.json;
+  };
+
+  buildCudaToolkitRedistPackage = callPackage ../development/compilers/cudatoolkit/redist/build-cuda-redist-package.nix { };
+
+  buildCudaToolkitRedistPackages = { version, manifest }: let
+    attrs = filterAttrs (key: value: key != "release_date") (importJSON manifest);
+  in mapAttrs buildCudaToolkitRedistPackage attrs;
+
+  cudaToolkitRedistPackages = if hasAttr cudaVersion cudaToolkitRedistManifests then
+    (buildCudaToolkitRedistPackages { version = cudaVersion; manifest = cudaToolkitRedistManifests.${cudaVersion}; })
+  else {};
+
+  ### Override functions
+  
+  addBuildInputs = drv: buildInputs: drv.overrideAttrs (oldAttrs: {
+    buildInputs = (oldAttrs.buildInputs or []) ++ buildInputs;
+  });
+
+in cudaToolkitRedistPackages // {
+
+  inherit addBuildInputs cudatoolkit pkgs;
+
+  ### Additional packages such as cudnn, cutensor and nccl
+
+  cudnn = buildCuDnnPackage cudnnVersions.${cudnnVersionForCuda};
+
+  cutensor = buildCuTensorPackage rec {
+    version = if cudaMajorMinorVersion == "10.1" then "1.2.2.5" else "1.3.1.3";
+    inherit (cuTensorVersions.${version}) sha256;
+    # This can go into generic.nix
+    libPath = "lib/${if cudaMajorVersion == "10" then cudaMajoMinorVersion else cudaMajorVersion}";
+  };
+
+  magma = pkgs.magma.override { inherit cudatoolkit; };
+
+  nccl = callPackage ../development/libraries/science/math/nccl { };
+
+})).overrideScope' (final: prev: {
+
+  ### Overrides to fix the components of cudatoolkit-redist
+
+  libcufile = prev.libcufile.overrideAttrs (oldAttrs: {
+    nativeBuildInputs = oldAttrs.nativeBuildInputs ++ [
+      pkgs.addOpenGLRunpath
+    ];
+    buildInputs = oldAttrs.buildInputs ++ [
+      prev.libcublas
+      pkgs.numactl
+      pkgs.rdma-core
+    ];
+    postFixup = ''
+      ln -s $out/lib/libcufile.so $out/lib/libcufile.so.0
+      addOpenGLRunpath --force-rpath {$out,$lib}/lib/lib*.so
+      for program in $out/tools/gdsio; do
+        isELF "$program" || continue
+        addOpenGLRunpath "$program"
+      done
+
+      # addOpenGLRunpath --force-rpath {$out}/tools/*
```

Comment:
```suggestion
```

I've found that this is unnecessary to build

Reply:
That's because `autoPatchelfIgnoreMissingDeps = true;`. I'm not sure what should happen with the runpath/rpath when using `addOpenGLRunpath` and am looking for example binaries.

Reply:
Ah, yes I see what you mean. After playing around with this code more today, my main concern is figuring out how to get `autoPatchelfHook` to do something somewhat sane with `libcuda.so.1` without resorting to `autoPatchelfIgnoreMissingDeps = true;`. That flag strikes me as too dangerous, as I myself discovered here. I created a discussion on discourse: https://discourse.nixos.org/t/have-autopatchelfhook-ignore-only-some-missing-dependencies/18465?u=samuela.

Some random things off the top of my head:
* Somehow add a "custom rule" that specifies `libcuda.so.1` to `/run/opengl-driver/lib` in `autoPatchelfHook`. Not sure if autoPatchelf gives us that ability or not.
* Create a fake package containing a fake `libcuda.so.1`, and then include a second hook that patchelf find-replaces the fake path with `/run/opengl-driver/lib`.

wdyt?

Reply:
Let's see what people reply to on that discussion, and let's see if we can have a look at say a tensorflow or pytorch linked to libcuda, what does ldd say?

Reply:
After poking around through some of the `autoPatchelfHook` internals, I'm not seeing any way to sneak something into their lookup table. So I guess we're stuck with # 2? Am I missing a cleaner solution... I hope?